src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp

```diff
@@ -0,0 +1,187 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "precompiled.hpp"
+#include "asm/macroAssembler.hpp"
+#include "classfile/javaClasses.inline.hpp"
+#include "classfile/symbolTable.hpp"
+#include "include/jvm.h"
+#include "jni.h"
+#include "logging/log.hpp"
+#include "logging/logStream.hpp"
+#include "memory/allocation.inline.hpp"
+#include "memory/resourceArea.hpp"
+#include "oops/arrayOop.inline.hpp"
+#include "prims/universalUpcallHandler.hpp"
+#include "runtime/interfaceSupport.inline.hpp"
+#include "runtime/javaCalls.hpp"
+#include "runtime/jniHandles.inline.hpp"
+
+static constexpr CodeBuffer::csize_t upcall_stub_size = 1024;
+
+extern struct JavaVM_ main_vm;
+
+static struct {
+  bool inited;
+  struct {
+    Klass* klass;
+    Symbol* name;
+    Symbol* sig;
+  } upcall_method;  // jdk.internal.foreign.abi.UniversalUpcallHandler::invoke
+} upcall_info;
+
+// FIXME: This should be initialized explicitly instead of lazily/racily
+static void upcall_init() {
+#if 0
+  fprintf(stderr, "upcall_init()\n");
+#endif
+
+  TRAPS = Thread::current();
+  ResourceMark rm;
+
+  const char* cname = "jdk/internal/foreign/abi/ProgrammableUpcallHandler";
+  const char* mname = "invoke";
+  const char* mdesc = "(Ljdk/internal/foreign/abi/ProgrammableUpcallHandler;J)V";
+  Symbol* cname_sym = SymbolTable::new_symbol(cname, (int)strlen(cname));
+  Symbol* mname_sym = SymbolTable::new_symbol(mname, (int)strlen(mname));
+  Symbol* mdesc_sym = SymbolTable::new_symbol(mdesc, (int)strlen(mdesc));
+
+#if 0
+  ::fprintf(stderr, "cname_sym: %p\n", cname_sym);
+  ::fprintf(stderr, "mname_sym: %p\n", mname_sym);
+  ::fprintf(stderr, "mdesc_sym: %p\n", mdesc_sym);
+#endif
+
+  Klass* k = SystemDictionary::resolve_or_null(cname_sym, THREAD);
+#if 0
+  ::fprintf(stderr, "Klass: %p\n", k);
+#endif
+
+  Method* method = k->lookup_method(mname_sym, mdesc_sym);
+#if 0
+  ::fprintf(stderr, "Method: %p\n", method);
+#endif
+
+  upcall_info.upcall_method.klass = k;
+  upcall_info.upcall_method.name = mname_sym;
+  upcall_info.upcall_method.sig = mdesc_sym;
+
+  upcall_info.inited = true;
+}
+
+static void upcall_helper(jobject rec, address buff) {
+  void *p_env = NULL;
+
+  Thread* thread = Thread::current_or_null();
+  if (thread == NULL) {
+    JavaVM_ *vm = (JavaVM *)(&main_vm);
+    vm -> functions -> AttachCurrentThreadAsDaemon(vm, &p_env, NULL);
```

Comment:
The context is a thread that is spawned by native code doing an upcall. We need to attach the thread to the VM first in that case. Normally this would be handled by the calling code, but in our case the calling code doesn't know it's calling into Java.

Reply:
Where's the logic for the native thread to detach? We have a similar problem in libgraal. We have a [utility class](https://github.com/oracle/graal/blob/e4b9ab931940e1946f96f2015b937ba100384573/compiler/src/org.graalvm.compiler.core/src/org/graalvm/compiler/core/GraalServiceThread.java#L27-L32) for libgraal created threads (as opposed to VM created threads that call into libgraal) that call into the VM. The utility class takes care of [attaching](https://github.com/oracle/graal/blob/a913944a06425c25ccd6e4a90379938fcf7ea2cf/substratevm/src/com.oracle.svm.graal.hotspot.libgraal/src/com/oracle/svm/graal/hotspot/libgraal/LibGraalFeature.java#L749) and [detaching](https://github.com/oracle/graal/blob/a913944a06425c25ccd6e4a90379938fcf7ea2cf/substratevm/src/com.oracle.svm.graal.hotspot.libgraal/src/com/oracle/svm/graal/hotspot/libgraal/LibGraalFeature.java#L757) to/from the VM.

Reply:
I added a call to DetechCurrentThread here: https://github.com/openjdk/jdk/pull/634/commits/719224ca9dc70fce6d28885acfb362fee715ebbd#diff-c084afc373a6ce95010a480ddc5ab79d3cb759b80e46102c212c2cbc948e2303R65

Reply:
Ok. That makes for high overhead for each upcall on a non-attached native thread. I assume that's an edge case not worth optimizing?